src/de/raw.rs

```diff
@@ -0,0 +1,905 @@
+                        nullbyte
+                    )));
+                }
+            }
+            std::cmp::Ordering::Greater => {
+                return Err(Error::custom(format!(
+                    "document has bytes remaining that were not visited: {}",
+                    length_remaining
+                )));
+            }
+            _ => (),
+        }
+        Ok(())
+    }
+
+    /// Read a string from the BSON.
+    fn parse_string(&mut self) -> Result<String> {
+        read_string(&mut self.bytes, false)
+    }
+
+    /// Construct a `DocumentAccess` and pass it into the provided closure, returning the
+    /// result of the closure if no other errors are encountered.
+    fn deserialize_document<F, O>(&mut self, f: F) -> Result<O>
+    where
+        F: FnOnce(DocumentAccess<'_, 'de>) -> Result<O>,
+    {
+        let mut length_remaining = read_i32(&mut self.bytes)? - 4;
+        let out = f(DocumentAccess {
+            root_deserializer: self,
+            length_remaining: &mut length_remaining,
+        });
+
+        if out.is_ok() {
+            self.end_document(length_remaining)?;
+        }
+        out
+    }
+}
+
+impl<'de, 'a> serde::de::Deserializer<'de> for &'a mut Deserializer<'de> {
+    type Error = Error;
+
+    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value>
+    where
+        V: serde::de::Visitor<'de>,
+    {
+        match self.current_type {
+            ElementType::Int32 => visitor.visit_i32(read_i32(&mut self.bytes)?),
+            ElementType::Int64 => visitor.visit_i64(read_i64(&mut self.bytes)?),
+            ElementType::Double => visitor.visit_f64(read_f64(&mut self.bytes)?),
+            ElementType::String => visitor.visit_string(self.parse_string()?),
+            ElementType::Boolean => visitor.visit_bool(read_bool(&mut self.bytes)?),
+            ElementType::Null => visitor.visit_unit(),
+            ElementType::ObjectId => {
+                let oid = ObjectId::from_reader(&mut self.bytes)?;
+                visitor.visit_map(ObjectIdAccess::new(oid))
+            }
+            ElementType::EmbeddedDocument => {
+                self.deserialize_document(|access| visitor.visit_map(access))
+            }
+            ElementType::Array => self.deserialize_document(|access| visitor.visit_seq(access)),
+            ElementType::Binary => {
+                let len = read_i32(&mut self.bytes)?;
+                if !(0..=MAX_BSON_SIZE).contains(&len) {
+                    return Err(Error::invalid_length(
+                        len as usize,
+                        &format!("binary length must be between 0 and {}", MAX_BSON_SIZE).as_str(),
+                    ));
+                }
+                let subtype = BinarySubtype::from(read_u8(&mut self.bytes)?);
+                match subtype {
+                    BinarySubtype::Generic => {
+                        visitor.visit_borrowed_bytes(self.bytes.read_slice(len as usize)?)
```

Comment:
>this means that generic Binary values will deserialize to &[u8] where all other subtypes will be a {"$binary: {"subType": ..., "base64": ...}} map structure, right?

Yep, this is correct.

>Which I guess means a data structure intended to be deserialized from raw BSON must conform to either the generic shape, or the everything-else shape, and can't accept both.

This isn't the case, though. The `Binary` struct, for example, can deserialize from either representation. This is true of `Bson` as well, which `Binary` uses under the hood for deserialization. Things that require bytes (e.g. `ByteBuf`) will only be able to deserialize from the generic subtype binary, however. `Uuid` via our serde helper will only be able to use the map one, since it requires a specific subtype.